### DISCUSSION TITLE
Is it possible to mock publicRuntimeConfig with Jest?

### DISCUSSION THREAD

**Discussion by mitestainer** [vercel/next.js#41308] [Help]
I have a component that uses a value coming from `getConfig().publicRuntimeConfig`, which is set to 'false' by default. On the Jest tests I need to test for both 'false' and 'true' cases, but I can't seem to find a way to mock `next/config` correctly. 

I tried this, as seen on many answers on StackOverflow:
```
jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {
    key: 'value'
  }
}));
```
but `publicRuntimeConfig` keeps the same values as the ones on `next.config.js`.

Yes, I have the `setupFilesAfterEnv: ['./jest.setup.js']` line on `jest.config.js`, as well as 
```
import { setConfig } from 'next/config'
import config from '../next.config'

setConfig(config)
```
in the Jest setup file.

I can access the values on the `.env` file and even set new values for it on the tests, but I can't work with `publicRuntimeConfig`. So I wonder: is it possible to mock `next/config` or am I skipping something? 🤔

**Comment by icyJoseph** [comment]
Sad to report that it works for me...

![Screenshot 2022-10-11 at 09 09 25](https://user-images.githubusercontent.com/21013447/195019837-9d1e1761-5350-4493-aecc-e37ff0aa58b7.png)

Which outputs:

![Screenshot 2022-10-11 at 09 09 54](https://user-images.githubusercontent.com/21013447/195019930-9773fa4b-d94e-439b-bcc4-ba1249d11be4.png)

Are you using `const nextJest = require('next/jest');`? Though that shouldn't really be an issue.

**Reply by mitestainer** [reply]
Thanks for replying!

Yes, it works if I place the mock before all the tests:
![image](https://user-images.githubusercontent.com/37314933/195107864-90a043d0-e25d-4d70-83cf-87db1af613f1.png)

But if I try to mock a value inside a test, I still get the original values:
![image](https://user-images.githubusercontent.com/37314933/195108050-0d5c8269-3cb6-41a9-b784-b981b11c6cc7.png)

Heard that I should use `jest.doMock` if I want to use a mock for a single test as `jest.mock` is hoisted to the top of the code, but it does nothing...
![image](https://user-images.githubusercontent.com/37314933/195108818-dbf8b7df-8dde-4e1b-ab9b-4d91b051aab6.png)

And yes, I'm using `const nextJest = require('next/jest');`

**Reply by icyJoseph** [reply]
Ah, for this kind of thing, sometimes it is better to mock the entire module via __mocks__. I think `doMock` works, but only if you import after you declare to mock...

```jts
    jest.doMock('next/config', () => {
        let config: Record<string, unknown> = {foo: 'bar'};
        return {
            __esModule: true,
            default: () => config,
            setConfig: (next: Record<string, unknown>) => {
                config = next;
            },
        };
    });

    await import('next/config').then((_getConfig) => console.log('mod', _getConfig.default()));
```

I'd mock this at `__mock__` if I ran out of patience:

```ts
// src/__mocks__/next/config.ts

import originalGetConfig, {setConfig as originalSetConfig} from 'next/config';

let local: Record<string, unknown> = {};

const getConfig = () => {
    return {...originalGetConfig(), ...local};
};

function override(next: Record<string, unknown>) {
    local = next;
}

function reset() {
    local = {};
}

export const setConfig = originalSetConfig;

getConfig.override = override;
getConfig.reset = reset;

export default getConfig;
```

And then I can control inside the test how the config should respond, through `override` and `reset`. I guess you could also mock it like this in the test file though. You might have to tell TS to shut up with `// @ts-expect-error <message>`

Configuring mocks is always annoying, and I reckon perhaps there's a better way to test what you want to safe guard?

**Reply by mitestainer** [reply]
Yeah, mocking the entire module would probably be too much for my needs. I guess a simple refactor to not use `publicRuntimeConfig` inside the component, and pass the value I need a prop instead, will do the trick.

Thank you for the explanation!
